lib/kafo/data_type.rb

```diff
@@ -0,0 +1,99 @@
+module Kafo
+  class DataType
+    def self.new_from_string(str)
+      keyword_re = /\A([\w:]+)(?:\[(.*)\])?\z/m.match(str)
+      if keyword_re
+        if (type = @keywords[keyword_re[1]])
+          args = if keyword_re[2]
+                   hash_re = keyword_re[2].match(/\A\s*{(.*)}\s*\z/m)
+                   if hash_re
+                     [parse_hash(hash_re[1])]
+                   else
+                     split_arguments(keyword_re[2])
+                   end
+                 else
+                   []
+                 end
+          type.new(*args)
+        elsif ['Data'].include?(keyword_re[1])
+          DataTypes::Any.new
+        elsif keyword_re[1] == 'Default'
+          DataTypes::Enum.new('default')
+        elsif @keywords[keyword_re[1].capitalize]  # pre-Puppet 4 data types
+          DataTypes::Optional.new(keyword_re[1].capitalize)
+        else
+          raise ConfigurationException, "unknown data type #{keyword_re[1]}"
+        end
+      else
+        raise ConfigurationException, "data type not recognized #{str}"
+      end
+    end
+
+    def self.register_type(keyword, type)
+      @keywords ||= {}
+      @keywords[keyword] = type
```

Comment:
it might raise exception if the type is registered already? now when hooks can come from installer plugins this can happen more often